<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jacob WIlliams" />

<meta name="date" content="2024-10-23" />

<title>RICE</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">RICE</h1>
<h4 class="author">Jacob WIlliams</h4>
<h4 class="date">2024-10-23</h4>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="do">## Load Relevant Packages</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(gdsfmt)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(SeqArray)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(SeqVarTools)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(STAAR)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg38.knownGene)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">library</span>(SCANG)</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="fu">library</span>(STAARpipeline)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="fu">library</span>(glmnet)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="fu">library</span>(caret)</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="fu">library</span>(SuperLearner)</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="fu">library</span>(RICE)</span></code></pre></div>
<p>RICE performs risk prediction for complex traits and diseases by
incorporating both common (RICE-CV) and rare variants (RICE-RV). RICE
framework consists of three main steps: (1) RICE-CV combines multiple
existing common variant PRSs using ensemble learning to generate a
single robust PRS; (2) RICE-RV identifies significant rare variants sets
conditioned on RICE-CV, models them using statistical methods, and
combines the results through ensemble learning to produce a rare variant
PRS; (3) The PRS from RICE-CV and RICE-RV are then evaluated together in
a regression model, adjusting for covariates. RICE requires three
independent datasets: (1) a training dataset to generate GWAS summary
statistics, rare variant association p-values, and train model for both
common and rare variant models; (2) a tuning dataset to optimize model
parameters and train the ensemble learning step; (3) a validation
dataset to assess the final prediction performance.</p>
<p>This vignette illustrates an example of constructing a rare variant
PRS using chromosome 22 of 1000 Genomes data and provides details on the
SuperLearner ensemble regression algorithm used in both RICE-RV and
RICE-CV. Data included in this vignette is located at <a href="https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/Y7AXYA" class="uri">https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/Y7AXYA</a>
and can be either directly downloaded from this link or using wget.</p>
<p>Note, this vignette shows how to create a rare variant PRS using
burden scores extracted from an AGDS file. However, if a user has burden
scores from another source, a rare variant PRS can still be constructed
following the RICE-RV structure. If this is the case, start at the
Penalized Regression section below.</p>
<div id="rice-rv" class="section level1">
<h1>RICE-RV</h1>
<p>RICE-RV performs rare variant association analysis to identify
significant rare variant sets using the training dataset with
STAARpipeline. With these significant rare variant sets, burden scores
are constructed for each set. Then penalized regression models are
trained using the burden scores and predicted onto the tuning and
validation datasets. Ensemble regression using SuperLearner optimizes
the predictions to form a single PRS. Lastly, this PRS is projected onto
the validation set to assess the final prediction performance.</p>
<p>The first step of RICE-RV involves STAARpipeline. RICE-RV can be seen
as an extension of the STAARpipeline, as it relies on the results of
STAARpipeline and operates with identical data structure. Therefore,
before attempting to implement RICE-RV, we suggest reading through the
STAARpipeline tutorial (<a href="https://github.com/xihaoli/STAARpipeline-Tutorial" class="uri">https://github.com/xihaoli/STAARpipeline-Tutorial</a>). This
vignette will only provide brief summaries of conducting rare variant
analysis with STAARpipeline.</p>
<div id="agds-files" class="section level2">
<h2>aGDS Files</h2>
<p>The first step of both RICE-RV is creating annotated genomic data
structure files (aGDS) for each chromosome. In the manuscript, aGDS
files included the train, tune, and validation sets, as we can extract
each set by subsetting. These are highly compressed files that include
all variants and the functional annotations of each variant. These are
commonly created using VCF files. Full details of creating aGDS files
can be found in the STAARpipeline tutorial (<a href="https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#generate-genomic-data-structure-gds-file" class="uri">https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#generate-genomic-data-structure-gds-file</a>).</p>
</div>
<div id="baseline-model" class="section level2">
<h2>Baseline Model</h2>
<p>The second step of both RICE-RV is creating a baseline null model to
conduct rare variant association analysis with the training dataset.
RICE-RV is conducted after RICE-CV therefore the baseline model for rare
variant association analysis includes the common variant PRS from
RICE-CV. In the manuscript, RICE-CV was constructed with CT, LDpred2,
and Lassosum2 in the case of the training dataset consisting of a single
ancestry and CT-SLEB, PROSPER, and JointPRS in the case of the training
dataset consisting of multiple ancestries. Full details on the
implementation of this can be found at â€¦ and <a href="https://github.com/jwilliams10/RareVariantPRS" class="uri">https://github.com/jwilliams10/RareVariantPRS</a>. However,
in practice RICE-RV can be constructed conditional on any common variant
PRS. Therefore, here we fit the baseline model with a simulated common
variant PRS.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="do">## Create Train/Tune/Validation Partitions</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1330</span>)</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="do">### Get Sample IDs</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>genofile <span class="ot">&lt;-</span> <span class="fu">seqOpen</span>(<span class="st">&quot;Chr22_1000G.gds&quot;</span>)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>sampleids <span class="ot">&lt;-</span> <span class="fu">seqGetData</span>(genofile,<span class="st">&quot;sample.id&quot;</span>)</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="fu">seqClose</span>(genofile)</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="do">### 70% Train, 15% Tune, 15% Validation</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>train_set <span class="ot">&lt;-</span> <span class="fu">sample</span>(sampleids,<span class="fu">round</span>(<span class="fu">length</span>(sampleids)<span class="sc">*</span><span class="fl">0.7</span>))</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>remaining_set <span class="ot">&lt;-</span> sampleids[<span class="sc">!</span>(sampleids <span class="sc">%in%</span> train_set)]</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>tune_set <span class="ot">&lt;-</span> <span class="fu">sample</span>(remaining_set,<span class="fu">round</span>(<span class="fu">length</span>(remaining_set)<span class="sc">*</span><span class="fl">0.5</span>))</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>validation_set <span class="ot">&lt;-</span> sampleids[<span class="sc">!</span>(sampleids <span class="sc">%in%</span> <span class="fu">c</span>(train_set,tune_set))]</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="do">## Fit Baseline Model for Training Data</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="do">### Load Phenotype and Common Variant PRS</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;Phenotype.RData&quot;</span>)</span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;Common_Variant_PRS.RData&quot;</span>)</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="do">### Merge Phenotype and Common Variant PRS</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>Phenotype_Data <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(Y,Common_Variant_PRS)</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="do">### Subset to the Training Set</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a>Phenotype_Data_Train <span class="ot">&lt;-</span> Phenotype_Data[Phenotype_Data<span class="sc">$</span>IDs <span class="sc">%in%</span> train_set,]</span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="do">### Fit Null Model without kinship, specifying continuous outcome, and column name of ID column</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a>Null_Model <span class="ot">&lt;-</span> <span class="fu">fit_nullmodel</span>(Y<span class="sc">~</span>PRS, <span class="at">data =</span> Phenotype_Data_Train,<span class="at">id =</span> <span class="st">&quot;IDs&quot;</span>,<span class="at">kins =</span> <span class="cn">NULL</span>,<span class="at">family =</span> <span class="fu">gaussian</span>(<span class="at">link =</span> <span class="st">&quot;identity&quot;</span>))</span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt; [1] &quot;kins is NULL, fit generalized linear model.&quot;</span></span></code></pre></div>
</div>
<div id="rare-variant-association-study" class="section level2">
<h2>Rare Variant Association Study</h2>
<p>The third step of RICE-RV is to perform rare variant association
analysis using STAARpipeline. Rare variant association analysis within
STAARpipeline can be divided into two analyses; gene-centric coding
analysis and gene-centric noncoding analysis. Gene-centric coding
analysis conducts association analysis with rare variants located in
protein-coding genes in the coding region and gene-centric noncoding
analysis in the noncoding region. Whole exome sequencing data is largely
comprised of rare variants located in the coding region while whole
genome sequencing data has rare variants in both the coding and
noncoding regions. This vignette only provides details for an example
gene-centric coding analysis. Full details on how to implement
gene-centric noncoding analysis and implementing both coding and
noncoding analysis within a HPC is available at <a href="https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#step-31-gene-centric-coding-analysis" class="uri">https://github.com/xihaoli/STAARpipeline-Tutorial?tab=readme-ov-file#step-31-gene-centric-coding-analysis</a>.</p>
<p>Here we implement gene-centric coding analysis on the training
dataset for a subset of chromosome 22. We implement for only a subset as
this is a non-optimal way of implementing this and takes a while to run,
if you have access to a HPC or cloud computing we recommend submitting
this as a batch job as shown in the STAARpipeline tutorial.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">### Gene-Centric Coding Analysis</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="do">## QC_label</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>QC_label <span class="ot">&lt;-</span> <span class="st">&quot;annotation/info/QC_label&quot;</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="do">## variant_type</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>variant_type <span class="ot">&lt;-</span> <span class="st">&quot;SNV&quot;</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="do">## geno_missing_imputation</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>geno_missing_imputation <span class="ot">&lt;-</span> <span class="st">&quot;mean&quot;</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="do">## Annotation_dir</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>Annotation_dir <span class="ot">&lt;-</span> <span class="st">&quot;annotation/info/FunctionalAnnotation&quot;</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="do">## Annotation channel</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>Annotation_name_catalog <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;Annotation_name_catalog.csv&quot;</span>)</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="do">## Use_annotation_weights</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>Use_annotation_weights <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="do">## Annotation name</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>Annotation_name <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;CADD&quot;</span>,<span class="st">&quot;LINSIGHT&quot;</span>,<span class="st">&quot;FATHMM.XF&quot;</span>,<span class="st">&quot;aPC.EpigeneticActive&quot;</span>,<span class="st">&quot;aPC.EpigeneticRepressed&quot;</span>,<span class="st">&quot;aPC.EpigeneticTranscription&quot;</span>,</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>                     <span class="st">&quot;aPC.Conservation&quot;</span>,<span class="st">&quot;aPC.LocalDiversity&quot;</span>,<span class="st">&quot;aPC.Mappability&quot;</span>,<span class="st">&quot;aPC.TF&quot;</span>,<span class="st">&quot;aPC.Protein&quot;</span>)</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="do">## Protein-Coding Genes in Chromosome 22</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>genes_info_chr <span class="ot">&lt;-</span> genes_info[genes_info[,<span class="dv">2</span>]<span class="sc">==</span><span class="dv">22</span>,]</span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>Genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;OR11H1&quot;</span>,<span class="st">&quot;POTEH&quot;</span>,<span class="st">&quot;CCT8L2&quot;</span>,<span class="st">&quot;XKR3&quot;</span>,<span class="st">&quot;GAB4&quot;</span>,<span class="st">&quot;IL17RA&quot;</span>,<span class="st">&quot;RANBP1&quot;</span>,<span class="st">&quot;PI4KA&quot;</span>,<span class="st">&quot;SERPIND1&quot;</span>,<span class="st">&quot;ZNF280B&quot;</span>,<span class="st">&quot;GGTLC2&quot;</span>,<span class="st">&quot;PIK3IP1&quot;</span>)</span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>genes_info_chr_subset <span class="ot">&lt;-</span> genes_info_chr[genes_info_chr<span class="sc">$</span>hgnc_symbol <span class="sc">%in%</span> Genes,]</span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="do">### Gene-Centric Coding Analysis</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a>genofile <span class="ot">&lt;-</span> <span class="fu">seqOpen</span>(<span class="st">&quot;Chr22_1000G.gds&quot;</span>)</span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a>results_coding <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a></span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="cf">for</span>(kk <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(genes_info_chr_subset)){</span>
<span id="cb3-30"><a href="#cb3-30" tabindex="-1"></a>  <span class="fu">print</span>(kk)</span>
<span id="cb3-31"><a href="#cb3-31" tabindex="-1"></a>  gene_name <span class="ot">&lt;-</span> genes_info_chr_subset[kk,<span class="dv">1</span>]</span>
<span id="cb3-32"><a href="#cb3-32" tabindex="-1"></a>  results <span class="ot">&lt;-</span> <span class="fu">Gene_Centric_Coding</span>(<span class="at">chr=</span><span class="dv">22</span>,<span class="at">gene_name=</span>gene_name,<span class="at">genofile=</span>genofile,<span class="at">obj_nullmodel=</span>Null_Model,</span>
<span id="cb3-33"><a href="#cb3-33" tabindex="-1"></a>                                 <span class="at">rare_maf_cutoff=</span><span class="fl">0.01</span>,<span class="at">rv_num_cutoff=</span><span class="dv">2</span>,</span>
<span id="cb3-34"><a href="#cb3-34" tabindex="-1"></a>                                 <span class="at">QC_label=</span>QC_label,<span class="at">variant_type=</span>variant_type,<span class="at">geno_missing_imputation=</span>geno_missing_imputation,</span>
<span id="cb3-35"><a href="#cb3-35" tabindex="-1"></a>                                 <span class="at">Annotation_dir=</span>Annotation_dir,<span class="at">Annotation_name_catalog=</span>Annotation_name_catalog,</span>
<span id="cb3-36"><a href="#cb3-36" tabindex="-1"></a>                                 <span class="at">Use_annotation_weights=</span>Use_annotation_weights,<span class="at">Annotation_name=</span>Annotation_name,<span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-37"><a href="#cb3-37" tabindex="-1"></a>  </span>
<span id="cb3-38"><a href="#cb3-38" tabindex="-1"></a>  results_coding <span class="ot">&lt;-</span> <span class="fu">append</span>(results_coding,results)</span>
<span id="cb3-39"><a href="#cb3-39" tabindex="-1"></a>}</span>
<span id="cb3-40"><a href="#cb3-40" tabindex="-1"></a><span class="fu">seqClose</span>(genofile)</span></code></pre></div>
</div>
<div id="burden-scores" class="section level2">
<h2>Burden Scores</h2>
<p>The fourth step of is extracting the burden scores of the significant
rare variant sets. In our manuscript, significance is defined as having
the STAAR-Burden p-value less than <span class="math inline">\(10^{-3}\)</span>, however, an individual can
construct burden scores for any set of rare variant sets. To identify
significant rare variant sets, we first compile the results from the
gene-centric coding analysis.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="do">### Compile Gene-Centric Coding Analysis</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>coding_sig <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(results_coding)){</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  coding_sig <span class="ot">&lt;-</span> <span class="fu">rbind</span>(coding_sig,<span class="fu">unlist</span>(results_coding[[j]][,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,<span class="dv">61</span>)]))</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>}</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>coding_sig <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(coding_sig)</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="fu">colnames</span>(coding_sig) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Gene&quot;</span>,<span class="st">&quot;Chr&quot;</span>,<span class="st">&quot;Category&quot;</span>,<span class="st">&quot;Number_SNV&quot;</span>,<span class="st">&quot;STAARB&quot;</span>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>coding_sig<span class="sc">$</span>Chr <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(coding_sig<span class="sc">$</span>Chr) </span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>coding_sig<span class="sc">$</span>Number_SNV <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(coding_sig<span class="sc">$</span>Number_SNV) </span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>coding_sig<span class="sc">$</span>STAARB <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(coding_sig<span class="sc">$</span>STAARB) </span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>coding_sig <span class="ot">&lt;-</span> coding_sig[coding_sig<span class="sc">$</span>Number_SNV <span class="sc">&lt;</span> <span class="dv">2000</span>,]</span></code></pre></div>
<p>Again this code is written for this vignette, when using a HPC or
cloud computing refer to the STAARpipeline tutorial to optimally compile
the results. Once compiled the structure of the results should look like
this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="do">### Head of Significant Rare Variant Set Results</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>coding_sig <span class="ot">&lt;-</span> coding_sig[coding_sig<span class="sc">$</span>STAARB <span class="sc">&lt;</span> <span class="fl">1e-3</span>,]</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">head</span>(coding_sig)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt;        Gene Chr   Category Number_SNV       STAARB</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt; 18   RANBP1  22   missense          5 4.566458e-05</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt; 23    PI4KA  22 synonymous         60 5.117859e-04</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt; 25 SERPIND1  22   missense         15 6.412555e-04</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt; 28  ZNF280B  22   missense          8 8.451143e-08</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt; 30   GGTLC2  22   missense         26 5.101352e-06</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; 33  PIK3IP1  22 synonymous          3 5.468301e-07</span></span></code></pre></div>
<p>Note, the results include the protein-coding gene, the chromosome,
and the functional category. These are essential elements to define a
rare variant set to build and extract a burden score. Below is how to
compute the burden scores for all significant rare variant sets for all
individuals. Burden scores are independent across individuals, therefore
they can be computed for all individuals then separated into train,
tune, and validation sets. Further note, that the STAAR-Burden(1,1)
p-value is the only p-value, this is because this the p-value used to
determine significance in the manuscript. However, the raw output of the
rare variant association analysis includes much more information
outlined in the STAARpipeline tutorial. With the significant rare
variant sets, we can extract and build the burden scores of the
significant rare variant sets.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">### Significant Rare Variant Set&#39;s Burden Scores</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>genofile <span class="ot">&lt;-</span> <span class="fu">seqOpen</span>(<span class="st">&quot;Chr22_1000G.gds&quot;</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>Sig_Burdens <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="do">### Loop over significant rare variant sets</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="do">### Using IDs = sampleids to build burden scores for all individuals.</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(coding_sig)){</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  chr <span class="ot">&lt;-</span> coding_sig<span class="sc">$</span>Chr[i]</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  gene_name <span class="ot">&lt;-</span> coding_sig<span class="sc">$</span>Gene[i]</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  category <span class="ot">&lt;-</span> coding_sig<span class="sc">$</span>Category[i]</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>  a <span class="ot">&lt;-</span> <span class="fu">Burden_Scores</span>(<span class="at">region =</span> <span class="st">&quot;Coding&quot;</span>,<span class="at">chr =</span> chr, <span class="at">gene_name =</span> gene_name,<span class="at">category =</span> category,</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>              <span class="at">genofile =</span> genofile,<span class="at">IDs =</span> sampleids,<span class="at">rare_maf_cutoff=</span><span class="fl">0.01</span>,<span class="at">rv_num_cutoff=</span><span class="dv">2</span>,</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>              <span class="at">QC_label=</span>QC_label,<span class="at">variant_type=</span>variant_type,<span class="at">geno_missing_imputation=</span>geno_missing_imputation,</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>              <span class="at">Annotation_dir=</span>Annotation_dir,<span class="at">Annotation_name_catalog=</span>Annotation_name_catalog,<span class="at">silent =</span> <span class="cn">TRUE</span>) </span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a> <span class="do">### Combine them column-wise</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>  Sig_Burdens <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Sig_Burdens,a)</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>}</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="fu">seqClose</span>(genofile)</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="do">### The burden scores align with the order of the sampleids</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a>IDs_Sig_Burdens <span class="ot">&lt;-</span> sampleids</span></code></pre></div>
</div>
<div id="penalized-regression" class="section level2">
<h2>Penalized Regression</h2>
<p>The fifth step of RICE-RV trains penalized regression models with the
trait as the outcome and the significant burden scores as an input. Once
these are trained, we predict the outcome for the tuning and validation.
Note we do not choose optimal hyper-parameters for the penalized
regression models, we simply use the entire sequence of
hyper-parameters.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## Penalized Regression</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="do">### Align with phenotype data</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">IDs =</span> IDs_Sig_Burdens,Sig_Burdens)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>Phenotype_Data <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(Phenotype_Data,X)</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt; Joining with `by = join_by(IDs)`</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="do">### Extract Aligned Burdens, change structure</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>X_Aligned <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(Phenotype_Data[,<span class="fu">paste0</span>(<span class="st">&quot;X&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(Sig_Burdens))])</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="do">### Subset the Burden Scores to Train, Tune, Validation</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>X_train <span class="ot">&lt;-</span> X_Aligned[Phenotype_Data<span class="sc">$</span>IDs <span class="sc">%in%</span> train_set,]</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>X_tune <span class="ot">&lt;-</span> X_Aligned[Phenotype_Data<span class="sc">$</span>IDs <span class="sc">%in%</span> tune_set,]</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>X_valid <span class="ot">&lt;-</span> X_Aligned[Phenotype_Data<span class="sc">$</span>IDs <span class="sc">%in%</span> validation_set,]</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="do">### Build a null model using control covariates to train penalized regression models</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>train_null_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(Y<span class="sc">~</span><span class="dv">1</span>,<span class="at">data =</span> Phenotype_Data[Phenotype_Data<span class="sc">$</span>IDs <span class="sc">%in%</span> train_set,])</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="do">### Extract residuals to use as a response</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>Residuals_Train <span class="ot">&lt;-</span> train_null_model<span class="sc">$</span>residuals</span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="do">### Train penalized regression models on the training set</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a>lasso_train <span class="ot">&lt;-</span> <span class="fu">glmnet</span>(X_train,Residuals_Train,<span class="at">family =</span> <span class="st">&quot;gaussian&quot;</span>,<span class="at">alpha =</span> <span class="dv">1</span>)</span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a>ridge_train <span class="ot">&lt;-</span> <span class="fu">glmnet</span>(X_train,Residuals_Train,<span class="at">family =</span> <span class="st">&quot;gaussian&quot;</span>,<span class="at">alpha =</span> <span class="dv">0</span>)</span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a><span class="do">### Project onto tuning and validation sets</span></span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>lasso_tune <span class="ot">&lt;-</span> <span class="fu">predict</span>(lasso_train,<span class="fu">as.matrix</span>(X_tune))</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a>ridge_tune <span class="ot">&lt;-</span> <span class="fu">predict</span>(ridge_train,X_tune)</span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a>lasso_valid <span class="ot">&lt;-</span> <span class="fu">predict</span>(lasso_train,X_valid)</span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a>ridge_valid <span class="ot">&lt;-</span> <span class="fu">predict</span>(ridge_train,X_valid)</span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a></span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a><span class="do">### Combine</span></span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a>all_tune <span class="ot">&lt;-</span> <span class="fu">cbind</span>(lasso_tune,ridge_tune)</span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a><span class="fu">colnames</span>(all_tune) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="st">&quot;lasso&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(lasso_tune)),<span class="fu">paste0</span>(<span class="st">&quot;ridge&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(ridge_tune)))</span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a>all_valid <span class="ot">&lt;-</span> <span class="fu">cbind</span>(lasso_valid,ridge_valid)</span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a><span class="fu">colnames</span>(all_valid) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="st">&quot;lasso&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(lasso_valid)),<span class="fu">paste0</span>(<span class="st">&quot;ridge&quot;</span>,<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(ridge_valid)))</span></code></pre></div>
</div>
<div id="superlearner" class="section level2">
<h2>SuperLearner</h2>
<p>The sixth and last step of RICE-RV, trains ensemble regression
SuperLearner algorithm on the predicted penalized regression outcomes
with the tuning dataset. The ensemble regression method used in RICE is
not simply the SuperLearner model, but an augmented algorithm of our own
design. The augmented ensemble regression uses SuperLearner to train an
ensemble regression algorithm which ensembles Lasso and ridge regression
algorithms. The performance of SuperLearner is compared against the
performance of the Lasso and ridge regression trained in SuperLearner.
The best performing prediction from the augmented SuperLearner based on
<span class="math inline">\(R^2\)</span> on the tuning data is then
compared to the prediction performance of the original penalized
regression outputs. Lastly, the best predictor on the tuning data is the
final rare variant PRS.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## SuperLearner</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="do">### Build a null model using control covariates to train the SuperLearner</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>tune_null_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(Y<span class="sc">~</span><span class="dv">1</span>,<span class="at">data =</span> Phenotype_Data[Phenotype_Data<span class="sc">$</span>IDs <span class="sc">%in%</span> tune_set,])</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>y_tune <span class="ot">&lt;-</span> tune_null_model<span class="sc">$</span>residuals</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">### Compute the R2 of the penalized regression inputs</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>best_r2 <span class="ot">&lt;-</span> <span class="fu">vector</span>()</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>count <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="fu">colnames</span>(all_tune)){</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">y =</span> y_tune,<span class="at">x =</span> all_tune[,j])</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>  best_r2[count] <span class="ot">&lt;-</span> <span class="fu">summary</span>(<span class="fu">lm</span>(y<span class="sc">~</span>x,<span class="at">data =</span> tmp))<span class="sc">$</span>r.squared</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>  count <span class="ot">&lt;-</span> count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>}</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="do">### Store the best</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>best_singleprs <span class="ot">&lt;-</span> <span class="fu">colnames</span>(all_tune)[<span class="fu">which.max</span>(best_r2)]</span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a>r2_best_singleprs_tune <span class="ot">&lt;-</span> best_r2[<span class="fu">which.max</span>(best_r2)]</span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="do">### Clean the projections</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a>mtx <span class="ot">&lt;-</span> <span class="fu">cor</span>(all_tune)</span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a>drop <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">as.data.frame</span>(all_tune))[<span class="fu">apply</span>(mtx,<span class="dv">2</span>,<span class="cf">function</span>(x){<span class="fu">sum</span>(<span class="fu">is.na</span>(x))}) <span class="sc">==</span> (<span class="fu">nrow</span>(mtx) <span class="sc">-</span> <span class="dv">1</span>)]</span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a>  </span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a>all_tune <span class="ot">&lt;-</span> all_tune[,<span class="sc">!</span>(<span class="fu">colnames</span>(all_tune) <span class="sc">%in%</span> drop)]</span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a>all_valid <span class="ot">&lt;-</span> all_valid[,<span class="sc">!</span>(<span class="fu">colnames</span>(all_valid) <span class="sc">%in%</span> drop)]</span>
<span id="cb8-25"><a href="#cb8-25" tabindex="-1"></a></span>
<span id="cb8-26"><a href="#cb8-26" tabindex="-1"></a>drop <span class="ot">&lt;-</span> <span class="fu">findLinearCombos</span>(all_tune)<span class="sc">$</span>remove</span>
<span id="cb8-27"><a href="#cb8-27" tabindex="-1"></a>drop <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">as.data.frame</span>(all_tune))[drop]</span>
<span id="cb8-28"><a href="#cb8-28" tabindex="-1"></a></span>
<span id="cb8-29"><a href="#cb8-29" tabindex="-1"></a>all_tune <span class="ot">&lt;-</span> all_tune[,<span class="sc">!</span>(<span class="fu">colnames</span>(all_tune) <span class="sc">%in%</span> drop)]</span>
<span id="cb8-30"><a href="#cb8-30" tabindex="-1"></a>all_valid <span class="ot">&lt;-</span> all_valid[,<span class="sc">!</span>(<span class="fu">colnames</span>(all_valid) <span class="sc">%in%</span> drop)]</span>
<span id="cb8-31"><a href="#cb8-31" tabindex="-1"></a></span>
<span id="cb8-32"><a href="#cb8-32" tabindex="-1"></a>mtx <span class="ot">&lt;-</span> <span class="fu">cor</span>(all_tune)</span>
<span id="cb8-33"><a href="#cb8-33" tabindex="-1"></a>drop <span class="ot">&lt;-</span> <span class="fu">findCorrelation</span>(mtx,<span class="at">cutoff=</span><span class="fl">0.98</span>)</span>
<span id="cb8-34"><a href="#cb8-34" tabindex="-1"></a>drop <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">as.data.frame</span>(all_tune))[drop]</span>
<span id="cb8-35"><a href="#cb8-35" tabindex="-1"></a>  </span>
<span id="cb8-36"><a href="#cb8-36" tabindex="-1"></a>all_tune <span class="ot">&lt;-</span> all_tune[,<span class="sc">!</span>(<span class="fu">colnames</span>(all_tune) <span class="sc">%in%</span> drop)]</span>
<span id="cb8-37"><a href="#cb8-37" tabindex="-1"></a>all_valid <span class="ot">&lt;-</span> all_valid[,<span class="sc">!</span>(<span class="fu">colnames</span>(all_valid) <span class="sc">%in%</span> drop)]</span>
<span id="cb8-38"><a href="#cb8-38" tabindex="-1"></a></span>
<span id="cb8-39"><a href="#cb8-39" tabindex="-1"></a><span class="do">### SuperLearner in RICE is an ensemble of Lasso and ridge regression models as we want to maintain linearity</span></span>
<span id="cb8-40"><a href="#cb8-40" tabindex="-1"></a>SL.library <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;SL.glmnet&quot;</span>,<span class="st">&quot;SL.ridge&quot;</span>)</span>
<span id="cb8-41"><a href="#cb8-41" tabindex="-1"></a>  </span>
<span id="cb8-42"><a href="#cb8-42" tabindex="-1"></a>full_superlearner <span class="ot">&lt;-</span> <span class="fu">SuperLearner</span>(<span class="at">Y =</span> y_tune, <span class="at">X =</span> <span class="fu">as.data.frame</span>(all_tune), <span class="at">family =</span> <span class="fu">gaussian</span>(),</span>
<span id="cb8-43"><a href="#cb8-43" tabindex="-1"></a>                                    <span class="at">SL.library =</span> SL.library,<span class="at">cvControl =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">20</span>))</span>
<span id="cb8-44"><a href="#cb8-44" tabindex="-1"></a><span class="do">### Perform cross validation of SuperLearner on the tuning dataset to see whether Lasso, Ridge, or SuperLearner performs best </span></span>
<span id="cb8-45"><a href="#cb8-45" tabindex="-1"></a>cvsl <span class="ot">&lt;-</span> <span class="fu">CV.SuperLearner</span>(<span class="at">Y =</span> y_tune, <span class="at">X =</span> <span class="fu">as.data.frame</span>(all_tune), <span class="at">family =</span> <span class="fu">gaussian</span>(),</span>
<span id="cb8-46"><a href="#cb8-46" tabindex="-1"></a>                          <span class="at">SL.library =</span> SL.library,<span class="at">cvControl =</span> <span class="fu">list</span>(<span class="at">V =</span> <span class="dv">20</span>))</span>
<span id="cb8-47"><a href="#cb8-47" tabindex="-1"></a></span>
<span id="cb8-48"><a href="#cb8-48" tabindex="-1"></a>best_algorithm <span class="ot">&lt;-</span> <span class="fu">summary</span>(cvsl)<span class="sc">$</span>Table<span class="sc">$</span>Algorithm[<span class="fu">which.min</span>(<span class="fu">summary</span>(cvsl)<span class="sc">$</span>Table<span class="sc">$</span>Ave)]</span>
<span id="cb8-49"><a href="#cb8-49" tabindex="-1"></a>  </span>
<span id="cb8-50"><a href="#cb8-50" tabindex="-1"></a><span class="do">### Predict all outcomes</span></span>
<span id="cb8-51"><a href="#cb8-51" tabindex="-1"></a>RareVariantPRS_Tune <span class="ot">&lt;-</span> <span class="fu">predict</span>(full_superlearner, <span class="fu">as.data.frame</span>(all_tune), <span class="at">onlySL =</span> <span class="cn">FALSE</span>)</span>
<span id="cb8-52"><a href="#cb8-52" tabindex="-1"></a>RareVariantPRS_Valid <span class="ot">&lt;-</span> <span class="fu">predict</span>(full_superlearner, all_valid, <span class="at">onlySL =</span> <span class="cn">FALSE</span>)</span>
<span id="cb8-53"><a href="#cb8-53" tabindex="-1"></a></span>
<span id="cb8-54"><a href="#cb8-54" tabindex="-1"></a>prs_best_tune_sl <span class="ot">&lt;-</span> RareVariantPRS_Tune<span class="sc">$</span>pred</span>
<span id="cb8-55"><a href="#cb8-55" tabindex="-1"></a>prs_best_tune_glmnet <span class="ot">&lt;-</span> RareVariantPRS_Tune<span class="sc">$</span>library.predict[,<span class="dv">1</span>]</span>
<span id="cb8-56"><a href="#cb8-56" tabindex="-1"></a>prs_best_tune_ridge <span class="ot">&lt;-</span> RareVariantPRS_Tune<span class="sc">$</span>library.predict[,<span class="dv">2</span>]</span>
<span id="cb8-57"><a href="#cb8-57" tabindex="-1"></a></span>
<span id="cb8-58"><a href="#cb8-58" tabindex="-1"></a>prs_best_valid_sl <span class="ot">&lt;-</span> RareVariantPRS_Valid<span class="sc">$</span>pred</span>
<span id="cb8-59"><a href="#cb8-59" tabindex="-1"></a>prs_best_valid_glmnet <span class="ot">&lt;-</span> RareVariantPRS_Valid<span class="sc">$</span>library.predict[,<span class="dv">1</span>]</span>
<span id="cb8-60"><a href="#cb8-60" tabindex="-1"></a>prs_best_valid_ridge <span class="ot">&lt;-</span> RareVariantPRS_Valid<span class="sc">$</span>library.predict[,<span class="dv">2</span>]</span>
<span id="cb8-61"><a href="#cb8-61" tabindex="-1"></a></span>
<span id="cb8-62"><a href="#cb8-62" tabindex="-1"></a><span class="do">### Choose best prediction from ensemble regression</span></span>
<span id="cb8-63"><a href="#cb8-63" tabindex="-1"></a><span class="cf">if</span>(best_algorithm <span class="sc">==</span> <span class="st">&quot;SL.glmnet_All&quot;</span>){</span>
<span id="cb8-64"><a href="#cb8-64" tabindex="-1"></a>    prs_best_tune <span class="ot">&lt;-</span> prs_best_tune_glmnet</span>
<span id="cb8-65"><a href="#cb8-65" tabindex="-1"></a>    prs_best_valid <span class="ot">&lt;-</span> prs_best_valid_glmnet</span>
<span id="cb8-66"><a href="#cb8-66" tabindex="-1"></a>}<span class="cf">else</span> <span class="cf">if</span>(best_algorithm <span class="sc">==</span> <span class="st">&quot;SL.ridge_All&quot;</span>){</span>
<span id="cb8-67"><a href="#cb8-67" tabindex="-1"></a>    prs_best_tune <span class="ot">&lt;-</span> prs_best_tune_ridge</span>
<span id="cb8-68"><a href="#cb8-68" tabindex="-1"></a>    prs_best_valid <span class="ot">&lt;-</span> prs_best_valid_ridge</span>
<span id="cb8-69"><a href="#cb8-69" tabindex="-1"></a>}<span class="cf">else</span> {</span>
<span id="cb8-70"><a href="#cb8-70" tabindex="-1"></a>    prs_best_tune <span class="ot">&lt;-</span> prs_best_tune_sl</span>
<span id="cb8-71"><a href="#cb8-71" tabindex="-1"></a>    prs_best_valid <span class="ot">&lt;-</span> prs_best_valid_sl</span>
<span id="cb8-72"><a href="#cb8-72" tabindex="-1"></a>}</span>
<span id="cb8-73"><a href="#cb8-73" tabindex="-1"></a></span>
<span id="cb8-74"><a href="#cb8-74" tabindex="-1"></a><span class="do">### Compute the R2 of the best ensemble regression prediction</span></span>
<span id="cb8-75"><a href="#cb8-75" tabindex="-1"></a>r2_sl_tune <span class="ot">&lt;-</span> <span class="fu">summary</span>(<span class="fu">lm</span>(y<span class="sc">~</span>x,<span class="at">data =</span> <span class="fu">data.frame</span>(<span class="at">y =</span> y_tune,<span class="at">x =</span> prs_best_tune)))<span class="sc">$</span>r.squared</span>
<span id="cb8-76"><a href="#cb8-76" tabindex="-1"></a></span>
<span id="cb8-77"><a href="#cb8-77" tabindex="-1"></a><span class="do">### Compare to the best individual input R2, keep the best</span></span>
<span id="cb8-78"><a href="#cb8-78" tabindex="-1"></a><span class="cf">if</span>(r2_sl_tune <span class="sc">&lt;</span> r2_best_singleprs_tune){</span>
<span id="cb8-79"><a href="#cb8-79" tabindex="-1"></a>  prs_best_tune <span class="ot">&lt;-</span> all_tune[,best_thresh]</span>
<span id="cb8-80"><a href="#cb8-80" tabindex="-1"></a>  prs_best_valid <span class="ot">&lt;-</span> all_valid[,best_thresh]</span>
<span id="cb8-81"><a href="#cb8-81" tabindex="-1"></a>}</span>
<span id="cb8-82"><a href="#cb8-82" tabindex="-1"></a></span>
<span id="cb8-83"><a href="#cb8-83" tabindex="-1"></a><span class="do">### Save final PRS for tuning and validation sets</span></span>
<span id="cb8-84"><a href="#cb8-84" tabindex="-1"></a>BestRareVariantPRS_Tune <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">IDs =</span> tune_set,<span class="at">PRS =</span> prs_best_tune)</span>
<span id="cb8-85"><a href="#cb8-85" tabindex="-1"></a>BestRareVariantPRS_Valid <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">IDs =</span> validation_set,<span class="at">PRS =</span> prs_best_valid)</span></code></pre></div>
<p>This ensemble regression algorithm is the exact same one used in
RICE-CV but the input is instead all of the PRSs from existing
methods.</p>
</div>
<div id="evaluation" class="section level2">
<h2>Evaluation</h2>
<p>We can evaluate the final prediction performance on the validation
dataset.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>valid_null_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(Y<span class="sc">~</span><span class="dv">1</span>,<span class="at">data =</span> Phenotype_Data[Phenotype_Data<span class="sc">$</span>IDs <span class="sc">%in%</span> validation_set,])</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>y_valid <span class="ot">&lt;-</span> valid_null_model<span class="sc">$</span>residuals</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="do">### R2 of RICE-RV on Validation Data</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(y<span class="sc">~</span>x,<span class="at">data =</span> <span class="fu">data.frame</span>(<span class="at">y=</span> y_valid,<span class="at">x =</span> prs_best_valid)))<span class="sc">$</span>r.squared</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt; [1] 0.05562963</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="do">### Beta of PRS per SD of RICE-RV on Validation Data</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="fu">coef</span>(<span class="fu">lm</span>(y<span class="sc">~</span>x <span class="sc">-</span> <span class="dv">1</span>,<span class="at">data =</span> <span class="fu">data.frame</span>(<span class="at">y=</span><span class="fu">scale</span>(y_valid),<span class="at">x =</span> <span class="fu">scale</span>(prs_best_valid))))</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt;         x </span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt; 0.2358593</span></span></code></pre></div>
</div>
<div id="extract-final-coefficients-of-rare-variant-sets" class="section level2">
<h2>Extract Final Coefficients of Rare Variant Sets</h2>
<p>Commonly, there is curiosity in knowing the relationship between the
variants and the final PRS. As all of the ensemble learning is done
using linear algorithms, the final predicted PRS is then a linear
projection of the significant burden scores. Using the tuning dataset,
we can estimate the final coefficients of the significant rare variant
sets with the following code.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="do">## Estimate Final Coefficients</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>Linear_Projection <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">y =</span> prs_best_tune,X_tune)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="do">### The tuning PRS is a perfect linear projection of the significant burden scores</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(y<span class="sc">~</span>.,<span class="at">data =</span>Linear_Projection))<span class="sc">$</span>r.squared</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co">#&gt; [1] 1</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="do">### Extract the coefficients</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>coding_sig<span class="sc">$</span>Final_Coefficients <span class="ot">&lt;-</span> <span class="fu">unname</span>(<span class="fu">coef</span>(<span class="fu">lm</span>(y<span class="sc">~</span>.,<span class="at">data =</span> Linear_Projection))[<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="fu">head</span>(coding_sig)</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co">#&gt;        Gene Chr   Category Number_SNV       STAARB Final_Coefficients</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co">#&gt; 18   RANBP1  22   missense          5 4.566458e-05         0.81881402</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co">#&gt; 23    PI4KA  22 synonymous         60 5.117859e-04         0.06437680</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a><span class="co">#&gt; 25 SERPIND1  22   missense         15 6.412555e-04         0.08098721</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="co">#&gt; 28  ZNF280B  22   missense          8 8.451143e-08         1.00794426</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co">#&gt; 30   GGTLC2  22   missense         26 5.101352e-06        -0.28491982</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co">#&gt; 33  PIK3IP1  22 synonymous          3 5.468301e-07         0.73066134</span></span></code></pre></div>
</div>
<div id="extensions" class="section level2">
<h2>Extensions</h2>
<p>Here we presented how to construct a rare variant PRS using rare
variants located in coding region of chromosome 22. This example is
almost identical to the simulation study in our manuscript but is far
from the complexities in real world data analyses in whole exome and
whole genome sequencing data. To extend this to real data analyses,
common extensions are the inclusion of all chromosomes, including both
rare variants located in coding and noncoding regions, and both binary
and continuous traits. To extend to all chromosomes, perform rare
variant association analysis on all chromosomes using the STAARpipeline,
extract burden scores for each chromosome using the code shown above,
compile the significant burden scores from all chromosomes together, and
perform the identical ensemble regression algorithm above. To extend to
noncoding regions, perform a separate gene-centric noncoding rare
variant association analysis with STAARpipeline, extract significant
burden scores using the same function but changing the region argument,
combine significant coding and noncoding burden scores, and perform the
identical ensemble regression algorithm. Lastly to extend to binary
traits, fit a binary null model with STAARpipeline, perform the
respective rare variant association tests, extract the significant
burden scores, implement the penalized regression models,and modify the
SuperLearner to use binary outcomes. These modifications include using
AUC computed in the <code>RISCA</code> package instead of <span class="math inline">\(R^2\)</span>, using
<code>method = &quot;method.AUC&quot;</code> in the SuperLearner call, and
dropping the ridge regression in <code>SL.library</code> as SuperLearner
does not support that for binary outcomes. The implementation of this
algorithm for binary traits for the UK Biobank whole exome sequencing
analysis can be found at <a href="https://github.com/jwilliams10/RareVariantPRS/blob/main/WES/Binary/RareVariant_PRS/Single_RareVariant_PRS_All.R" class="uri">https://github.com/jwilliams10/RareVariantPRS/blob/main/WES/Binary/RareVariant_PRS/Single_RareVariant_PRS_All.R</a>.</p>
</div>
<div id="sessioninfo" class="section level2">
<h2>SessionInfo</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co">#&gt; Running under: macOS Sonoma 14.7</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co">#&gt; time zone: America/New_York</span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="co">#&gt; [1] splines   stats4    stats     graphics  grDevices utils     datasets </span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="co">#&gt; [8] methods   base     </span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a><span class="co">#&gt;  [1] RICE_0.1.0                              </span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a><span class="co">#&gt;  [2] SuperLearner_2.0-29                     </span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a><span class="co">#&gt;  [3] gam_1.22-5                              </span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a><span class="co">#&gt;  [4] foreach_1.5.2                           </span></span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a><span class="co">#&gt;  [5] nnls_1.5                                </span></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a><span class="co">#&gt;  [6] caret_6.0-94                            </span></span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a><span class="co">#&gt;  [7] lattice_0.22-6                          </span></span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a><span class="co">#&gt;  [8] ggplot2_3.5.1                           </span></span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a><span class="co">#&gt;  [9] glmnet_4.1-8                            </span></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a><span class="co">#&gt; [10] STAARpipeline_0.9.7                     </span></span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a><span class="co">#&gt; [11] SCANG_1.0.4                             </span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a><span class="co">#&gt; [12] Matrix_1.7-0                            </span></span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a><span class="co">#&gt; [13] TxDb.Hsapiens.UCSC.hg38.knownGene_3.18.0</span></span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a><span class="co">#&gt; [14] GenomicFeatures_1.56.0                  </span></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a><span class="co">#&gt; [15] AnnotationDbi_1.66.0                    </span></span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a><span class="co">#&gt; [16] Biobase_2.64.0                          </span></span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a><span class="co">#&gt; [17] GenomicRanges_1.56.1                    </span></span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a><span class="co">#&gt; [18] GenomeInfoDb_1.40.1                     </span></span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a><span class="co">#&gt; [19] IRanges_2.38.1                          </span></span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a><span class="co">#&gt; [20] S4Vectors_0.42.1                        </span></span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a><span class="co">#&gt; [21] BiocGenerics_0.50.0                     </span></span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a><span class="co">#&gt; [22] STAAR_0.9.7                             </span></span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a><span class="co">#&gt; [23] dplyr_1.1.4                             </span></span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a><span class="co">#&gt; [24] SeqVarTools_1.42.0                      </span></span>
<span id="cb11-45"><a href="#cb11-45" tabindex="-1"></a><span class="co">#&gt; [25] SeqArray_1.44.0                         </span></span>
<span id="cb11-46"><a href="#cb11-46" tabindex="-1"></a><span class="co">#&gt; [26] gdsfmt_1.40.0                           </span></span>
<span id="cb11-47"><a href="#cb11-47" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-48"><a href="#cb11-48" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb11-49"><a href="#cb11-49" tabindex="-1"></a><span class="co">#&gt;   [1] BiocIO_1.14.0               bitops_1.0-8               </span></span>
<span id="cb11-50"><a href="#cb11-50" tabindex="-1"></a><span class="co">#&gt;   [3] tibble_3.2.1                GWASExactHW_1.2            </span></span>
<span id="cb11-51"><a href="#cb11-51" tabindex="-1"></a><span class="co">#&gt;   [5] GENESIS_2.34.0              hardhat_1.4.0              </span></span>
<span id="cb11-52"><a href="#cb11-52" tabindex="-1"></a><span class="co">#&gt;   [7] pROC_1.18.5                 XML_3.99-0.17              </span></span>
<span id="cb11-53"><a href="#cb11-53" tabindex="-1"></a><span class="co">#&gt;   [9] rpart_4.1.23                lifecycle_1.0.4            </span></span>
<span id="cb11-54"><a href="#cb11-54" tabindex="-1"></a><span class="co">#&gt;  [11] formula.tools_1.7.1         globals_0.16.3             </span></span>
<span id="cb11-55"><a href="#cb11-55" tabindex="-1"></a><span class="co">#&gt;  [13] MASS_7.3-61                 backports_1.5.0            </span></span>
<span id="cb11-56"><a href="#cb11-56" tabindex="-1"></a><span class="co">#&gt;  [15] magrittr_2.0.3              sass_0.4.9                 </span></span>
<span id="cb11-57"><a href="#cb11-57" tabindex="-1"></a><span class="co">#&gt;  [17] rmarkdown_2.28              jquerylib_0.1.4            </span></span>
<span id="cb11-58"><a href="#cb11-58" tabindex="-1"></a><span class="co">#&gt;  [19] yaml_2.3.10                 DBI_1.2.3                  </span></span>
<span id="cb11-59"><a href="#cb11-59" tabindex="-1"></a><span class="co">#&gt;  [21] minqa_1.2.8                 lubridate_1.9.3            </span></span>
<span id="cb11-60"><a href="#cb11-60" tabindex="-1"></a><span class="co">#&gt;  [23] abind_1.4-8                 zlibbioc_1.50.0            </span></span>
<span id="cb11-61"><a href="#cb11-61" tabindex="-1"></a><span class="co">#&gt;  [25] quadprog_1.5-8              purrr_1.0.2                </span></span>
<span id="cb11-62"><a href="#cb11-62" tabindex="-1"></a><span class="co">#&gt;  [27] RCurl_1.98-1.16             nnet_7.3-19                </span></span>
<span id="cb11-63"><a href="#cb11-63" tabindex="-1"></a><span class="co">#&gt;  [29] sandwich_3.1-1              ipred_0.9-15               </span></span>
<span id="cb11-64"><a href="#cb11-64" tabindex="-1"></a><span class="co">#&gt;  [31] lava_1.8.0                  GenomeInfoDbData_1.2.12    </span></span>
<span id="cb11-65"><a href="#cb11-65" tabindex="-1"></a><span class="co">#&gt;  [33] listenv_0.9.1               GWASTools_1.50.0           </span></span>
<span id="cb11-66"><a href="#cb11-66" tabindex="-1"></a><span class="co">#&gt;  [35] MatrixModels_0.5-3          parallelly_1.38.0          </span></span>
<span id="cb11-67"><a href="#cb11-67" tabindex="-1"></a><span class="co">#&gt;  [37] codetools_0.2-20            DelayedArray_0.30.1        </span></span>
<span id="cb11-68"><a href="#cb11-68" tabindex="-1"></a><span class="co">#&gt;  [39] DNAcopy_1.78.0              tidyselect_1.2.1           </span></span>
<span id="cb11-69"><a href="#cb11-69" tabindex="-1"></a><span class="co">#&gt;  [41] shape_1.4.6.1               UCSC.utils_1.0.0           </span></span>
<span id="cb11-70"><a href="#cb11-70" tabindex="-1"></a><span class="co">#&gt;  [43] lme4_1.1-35.5               matrixStats_1.4.1          </span></span>
<span id="cb11-71"><a href="#cb11-71" tabindex="-1"></a><span class="co">#&gt;  [45] GenomicAlignments_1.40.0    jsonlite_1.8.9             </span></span>
<span id="cb11-72"><a href="#cb11-72" tabindex="-1"></a><span class="co">#&gt;  [47] GMMAT_1.4.2                 mitml_0.4-5                </span></span>
<span id="cb11-73"><a href="#cb11-73" tabindex="-1"></a><span class="co">#&gt;  [49] survival_3.7-0              iterators_1.0.14           </span></span>
<span id="cb11-74"><a href="#cb11-74" tabindex="-1"></a><span class="co">#&gt;  [51] emmeans_1.10.4              tools_4.4.1                </span></span>
<span id="cb11-75"><a href="#cb11-75" tabindex="-1"></a><span class="co">#&gt;  [53] SNPRelate_1.38.0            Rcpp_1.0.13                </span></span>
<span id="cb11-76"><a href="#cb11-76" tabindex="-1"></a><span class="co">#&gt;  [55] glue_1.7.0                  prodlim_2024.06.25         </span></span>
<span id="cb11-77"><a href="#cb11-77" tabindex="-1"></a><span class="co">#&gt;  [57] pan_1.9                     SparseArray_1.4.8          </span></span>
<span id="cb11-78"><a href="#cb11-78" tabindex="-1"></a><span class="co">#&gt;  [59] xfun_0.47                   mgcv_1.9-1                 </span></span>
<span id="cb11-79"><a href="#cb11-79" tabindex="-1"></a><span class="co">#&gt;  [61] logistf_1.26.0              MatrixGenerics_1.16.0      </span></span>
<span id="cb11-80"><a href="#cb11-80" tabindex="-1"></a><span class="co">#&gt;  [63] withr_3.0.1                 fastmap_1.2.0              </span></span>
<span id="cb11-81"><a href="#cb11-81" tabindex="-1"></a><span class="co">#&gt;  [65] boot_1.3-31                 MultiSTAAR_0.9.7           </span></span>
<span id="cb11-82"><a href="#cb11-82" tabindex="-1"></a><span class="co">#&gt;  [67] fansi_1.0.6                 SparseM_1.84-2             </span></span>
<span id="cb11-83"><a href="#cb11-83" tabindex="-1"></a><span class="co">#&gt;  [69] digest_0.6.37               timechange_0.3.0           </span></span>
<span id="cb11-84"><a href="#cb11-84" tabindex="-1"></a><span class="co">#&gt;  [71] R6_2.5.1                    estimability_1.5.1         </span></span>
<span id="cb11-85"><a href="#cb11-85" tabindex="-1"></a><span class="co">#&gt;  [73] mice_3.16.0                 colorspace_2.1-1           </span></span>
<span id="cb11-86"><a href="#cb11-86" tabindex="-1"></a><span class="co">#&gt;  [75] RSQLite_2.3.7               utf8_1.2.4                 </span></span>
<span id="cb11-87"><a href="#cb11-87" tabindex="-1"></a><span class="co">#&gt;  [77] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span id="cb11-88"><a href="#cb11-88" tabindex="-1"></a><span class="co">#&gt;  [79] data.table_1.16.0           recipes_1.1.0              </span></span>
<span id="cb11-89"><a href="#cb11-89" tabindex="-1"></a><span class="co">#&gt;  [81] rtracklayer_1.64.0          class_7.3-22               </span></span>
<span id="cb11-90"><a href="#cb11-90" tabindex="-1"></a><span class="co">#&gt;  [83] httr_1.4.7                  S4Arrays_1.4.1             </span></span>
<span id="cb11-91"><a href="#cb11-91" tabindex="-1"></a><span class="co">#&gt;  [85] ModelMetrics_1.2.2.2        pkgconfig_2.0.3            </span></span>
<span id="cb11-92"><a href="#cb11-92" tabindex="-1"></a><span class="co">#&gt;  [87] gtable_0.3.5                timeDate_4041.110          </span></span>
<span id="cb11-93"><a href="#cb11-93" tabindex="-1"></a><span class="co">#&gt;  [89] blob_1.2.4                  lmtest_0.9-40              </span></span>
<span id="cb11-94"><a href="#cb11-94" tabindex="-1"></a><span class="co">#&gt;  [91] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span id="cb11-95"><a href="#cb11-95" tabindex="-1"></a><span class="co">#&gt;  [93] scales_1.3.0                png_0.1-8                  </span></span>
<span id="cb11-96"><a href="#cb11-96" tabindex="-1"></a><span class="co">#&gt;  [95] gower_1.0.1                 knitr_1.48                 </span></span>
<span id="cb11-97"><a href="#cb11-97" tabindex="-1"></a><span class="co">#&gt;  [97] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span id="cb11-98"><a href="#cb11-98" tabindex="-1"></a><span class="co">#&gt;  [99] rjson_0.2.23                nlme_3.1-166               </span></span>
<span id="cb11-99"><a href="#cb11-99" tabindex="-1"></a><span class="co">#&gt; [101] curl_5.2.3                  nloptr_2.1.1               </span></span>
<span id="cb11-100"><a href="#cb11-100" tabindex="-1"></a><span class="co">#&gt; [103] cachem_1.1.0                zoo_1.8-12                 </span></span>
<span id="cb11-101"><a href="#cb11-101" tabindex="-1"></a><span class="co">#&gt; [105] stringr_1.5.1               operator.tools_1.6.3       </span></span>
<span id="cb11-102"><a href="#cb11-102" tabindex="-1"></a><span class="co">#&gt; [107] parallel_4.4.1              restfulr_0.0.15            </span></span>
<span id="cb11-103"><a href="#cb11-103" tabindex="-1"></a><span class="co">#&gt; [109] pillar_1.9.0                grid_4.4.1                 </span></span>
<span id="cb11-104"><a href="#cb11-104" tabindex="-1"></a><span class="co">#&gt; [111] vctrs_0.6.5                 jomo_2.7-6                 </span></span>
<span id="cb11-105"><a href="#cb11-105" tabindex="-1"></a><span class="co">#&gt; [113] xtable_1.8-4                evaluate_1.0.0             </span></span>
<span id="cb11-106"><a href="#cb11-106" tabindex="-1"></a><span class="co">#&gt; [115] mvtnorm_1.3-1               cli_3.6.3                  </span></span>
<span id="cb11-107"><a href="#cb11-107" tabindex="-1"></a><span class="co">#&gt; [117] compiler_4.4.1              Rsamtools_2.20.0           </span></span>
<span id="cb11-108"><a href="#cb11-108" tabindex="-1"></a><span class="co">#&gt; [119] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span id="cb11-109"><a href="#cb11-109" tabindex="-1"></a><span class="co">#&gt; [121] future.apply_1.11.2         kinship2_1.9.6.1           </span></span>
<span id="cb11-110"><a href="#cb11-110" tabindex="-1"></a><span class="co">#&gt; [123] plyr_1.8.9                  stringi_1.8.4              </span></span>
<span id="cb11-111"><a href="#cb11-111" tabindex="-1"></a><span class="co">#&gt; [125] BiocParallel_1.38.0         munsell_0.5.1              </span></span>
<span id="cb11-112"><a href="#cb11-112" tabindex="-1"></a><span class="co">#&gt; [127] Biostrings_2.72.1           CompQuadForm_1.4.3         </span></span>
<span id="cb11-113"><a href="#cb11-113" tabindex="-1"></a><span class="co">#&gt; [129] quantreg_5.98               bit64_4.5.2                </span></span>
<span id="cb11-114"><a href="#cb11-114" tabindex="-1"></a><span class="co">#&gt; [131] future_1.34.0               quantsmooth_1.70.0         </span></span>
<span id="cb11-115"><a href="#cb11-115" tabindex="-1"></a><span class="co">#&gt; [133] KEGGREST_1.44.1             SummarizedExperiment_1.34.0</span></span>
<span id="cb11-116"><a href="#cb11-116" tabindex="-1"></a><span class="co">#&gt; [135] broom_1.0.7                 memoise_2.0.1              </span></span>
<span id="cb11-117"><a href="#cb11-117" tabindex="-1"></a><span class="co">#&gt; [137] bslib_0.8.0                 bit_4.5.0</span></span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
